### PR TITLE
[Breaking Change] Template helpers are now pure functions

### PR DESCRIPTION
**Pull request by naps62**
Changing the API of template helpers to always receive `@env` as first argument. Reasons below

This was made after a discussion with @gabrielpoca , which came from #123 

The original issue description felt "wrong" to me in a few ways, and would mean adding more meta magic into a module that already seemed to have a bit too much of it
Even though doing `include("file.css")` instead of `include(@env, "file.css")` within templates feels more comfortable, the magic required to do that seems counter to the usual way Elixir is done. And on a closer look, it actually removes from the template the usual explicitness that comes with Elixir. Those functions were not pure

The original issue would, in my opinion, just increase this problem, because it pushed to allow those helper functions to be called outside templates (by explicitly pasing `@env`) but still keep the original implicit api inside templates. This would make things inconsistent to look at, as well as making the codebase more complicated

Also, from what I understood there were issues with compilation times due to the fact that `use TemplateHelpers` had to be done for every single module (one per template file, I believe?). Changing this to `import TemplateHelpers` should also improve on that

### CODE REVIEW

**Comment by gabrielpoca** [comment]
I agree with this change. At first, I liked the idea of not having to pass the `env` around, but once I realized that I needed those functions outside the templates, I also understood that having two APIs for the same functions was confusing. I'm also in favor of removing features and keeping the codebase as simpler and smaller as possible. Still, I would like to hear from @frm on this one

**Comment by frm** [comment]
I have many mixed feelings on this. I prefer the explicitness but I also enjoy the comfort of not having to pass the `@env` variable. However we already see this in Phoenix with `@conn`.

The thing that annoys me the most is the API not being consistent. If we move this forward, I'd prefer to actually pass the `@env` to every Still template helper.

**Comment by naps62** [comment]
I agree with that last part (having `@env` as the first argument of all helpers). it keeps the elixir spirit, and prevents the caller from having to know if the specific function he's calling needs it or not

File: lib/still/preprocessor/renderer.ex [hunk 1]
```diff
@@ -80,7 +81,9 @@ defmodule Still.Preprocessor.Renderer do
             unquote(user_template_helpers_asts())
             unquote(renderer_ast)
 
-            use Still.Compiler.TemplateHelpers, unquote(Macro.escape(module_metadata))
+            import Still.Compiler.TemplateHelpers
+
+            Module.put_attribute(__MODULE__, :env, 1)
```
Comment (frm) [inline lib/still/preprocessor/renderer.ex:86]: 👍

Comment (naps62) [reply lib/still/preprocessor/renderer.ex]: fixed

**Comment by gabrielpoca** [comment]
I leave this up for you two to decide, but I don't relate to the premise that if `env` is always mandatory, the developer doesn't need to look it up. In Phoenix, you have the `form_for`, which doesn't receive `conn` because it doesn't need it. I'm not saying that Phoenix got everything right, but I also don't remember seeing any conversation where this was pointed out as an issue.

**Comment by naps62** [comment]
yep I have now revised my opinion @gabrielpoca xD

When I first thought about this, I had a different picture of the list of helpers we had here. I retract my original opinion :smile:

**Comment by frm** [comment]
I think it's due to a logical division. I agree that it feels arbitrary, though. String manipulation helpers wouldn't receive the `env` as it doesn't make sense. But everything else related to files would? What if they are file system related functions?

This is making me question the decision a lot. There are way too many arbitrary lines. The API still bothers me due to the lack of consistency, but I prefer explicitness rather than the magic we had before.

In summary, pure functions enable us to call the helpers directly so we gain in terms of integration and developer experience. If the downside is an inconsistent API, I'm fine with that.

Let's move it forward.

**Comment by frm** [comment]
@naps62 can you revert your last couple of commits?

**Comment by naps62** [comment]
@frm already did. I don't agree this makes the API inconsistent, to be honest. it still makes sense. it was the original argument "making it predicatble to the caller" that was a bit flawed, because applying that to things like a "truncate" function makes little sense

We can have a future discussion about whether it makes sense for all helpers to live in the same module. but regardless of the answer to that, I'm fine with the current approach for the time being

**Review by frm — APPROVED** [review]
